### PR TITLE
🐛 N°4170 - Fix encoding issue for long subjects (and other MIME UTF8 encoded data on multiple lines)

### PR DESCRIPTION
**Pull request by jbostoen**
Subject headers which were on multiple lines (and perhaps other headers too) suffer from a MIME decoding issue, causing whitespace or tabs to be added.

Below are two examples of subjects and how Outlook encoded the header.

FW: ⚠ This is a test with an emoji in the subject and a long subject message which will cause multi line subjects and encoding
```
Subject:
 =?utf-8?B?Rlc6IOKaoCBUaGlzIGlzIGEgdGVzdCB3aXRoIGFuIGVtb2ppIGluIHRoZSBz?=
 =?utf-8?B?dWJqZWN0IGFuZCBhIGxvbmcgc3ViamVjdCBtZXNzYWdlIHdoaWNoIHdpbGwg?=
 =?utf-8?Q?cause_multi_line_subjects_and_encoding?=
```

FW: ⚠ Aanpassing: 'verwijderde items' in Outlook zal na 90 dagen automatisch gewist worden

```
Subject:
 =?utf-8?B?Rlc6IOKaoCBBYW5wYXNzaW5nOiAndmVyd2lqZGVyZGUgaXRlbXMnIGluIE91?=
 =?utf-8?Q?tlook_zal_na_90_dagen_automatisch_gewist_worden?=
```

Debugging (logging the input and output of the DecodeHeaderString() method):
```
input: 	=?utf-8?B?Rlc6IOKaoCBUaGlzIGlzIGEgdGVzdCB3aXRoIGFuIGVtb2ppIGluIHRoZSBz?= =?utf-8?B?dWJqZWN0IGFuZCBhIGxvbmcgc3ViamVjdCBtZXNzYWdlIHdoaWNoIHdpbGwg?= =?utf-8?Q?cause_multi_line_subjects_and_encoding?=
output: 	FW: ⚠ This is a test with an emoji in the subject and a long subject message which will cause multi line subjects and encoding
input: =?utf-8?B?4pqgIFRoaXMgaXMgYSB0ZXN0IHdpdGggYW4gZW1vamkgaW4gdGhlIHN1Ympl?==?utf-8?B?Y3QgYW5kIGEgbG9uZyBzdWJqZWN0IG1lc3NhZ2Ugd2hpY2ggd2lsbCBjYXVz?==?utf-8?Q?e_multi_line_subjects_and_encoding?=
output: ⚠ This is a test with an emoji in the subject and a long subject message which will cause multi line subjects and encoding
```
This patch solves the MIME decoding by checking if there's still redundant leading white space (space, tab) before encoded strings. It also removes space in between encoded strings to be on the safe side.

### CODE REVIEW

**Comment by Molkobain** [comment]
Hello Jeffrey,

_This PR / bug has just been referenced as N°4170 on our bug tracker._

Thanks for the PR! Is this related to #7 where the subject was on several lines as well? If so can't we factorize the fact that some data may be on several lines?

Can you confirm that both examples are the same symptom of the **same** problem?

In order for us to move this PR forward, we would need some additions:
* Add 1 of the .eml (not .msg) files corresponding to these emails, named `email_133_kb4170_multiple_lines_encoded_data.eml` (check [here](https://github.com/Combodo/combodo-email-synchro/tree/master/test/emailsSample).
* Add a unit test to cover that case. You can take [this](https://github.com/Combodo/combodo-email-synchro/commit/e2b6ec6df6d809524edb422d9f1671d450478387) as an example, it's not perfect but it's better than nothing.

_Note: If you are not familiar with setting up a PHPUnit environment for your iTop, let us know, we'll try to help._

**Comment by jbostoen** [comment]
It's related. I assume the space(s) kept to fix non-encoded strings might have resulted in this behavior.

Can I add two?

**Comment by Molkobain** [comment]
> Can I add two?

Two unit tests? Yes!
Two .eml files? If they are 2 different use cases, sure.

**Comment by jbostoen** [comment]
Weird, in the .MSG file, the headers are

```
Subject:
 =?utf-8?B?Rlc6IOKaoCBUaGlzIGlzIGEgdGVzdCB3aXRoIGFuIGVtb2ppIGluIHRoZSBz?=
 =?utf-8?B?dWJqZWN0IGFuZCBhIGxvbmcgc3ViamVjdCBtZXNzYWdlIHdoaWNoIHdpbGwg?=
 =?utf-8?Q?cause_multi_line_subjects_and_encoding?=
```

I used ZamZar to convert it to an email file.
Now it's
```
Subject: =?utf-8?B?Rlc6IOKaoCBUaGlzIGlzIGEgdGVzdCB3aXRoIGFuIGVtb2ppIGluIHRoZSA=?=
	=?utf-8?B?c3ViamVjdCBhbmQgYSBsb25nIHN1YmplY3QgbWVzc2FnZSB3aGljaCB3?=
	=?utf-8?B?aWxsIGNhdXNlIG11bHRpIGxpbmUgc3ViamVjdHMgYW5kIGVuY29kaW5n?=
```

Basically it could still serve as a demo, but you can see the encoding has already changed a bit.
Does this still qualify for a test email? (definitely note the beginning of each third encrypted string!)

[emailsSample.zip](https://github.com/Combodo/combodo-email-synchro/files/6834491/emailsSample.zip)

**Comment by Molkobain** [comment]
You can locally test the .eml to see how they are decoded before your fix with [this script](https://github.com/Combodo/combodo-email-synchro/blob/master/utils/decoding-test.php), this way you can make sure the .eml files will still be good test samples.

Side note: Can't you download the message as .eml directly from your Exchange server?

File: test/emailsSample/readme.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
 email_131.eml: Email with "us-ascii" charset and binary attachment. (Issue was that attachment was converted through iconv() when it should not)
-email_132.eml: Email with uppercase content-transfer-encoding ("BASE64"). (Issue was that mail had "Quote-Printable" and "BASE64" instead of lowercase ones)
\ No newline at end of file
+email_132.eml: Email with uppercase content-transfer-encoding ("BASE64"). (Issue was that mail had "Quote-Printable" and "BASE64" instead of lowercase ones)
+email_133.eml: Email with a multiline subject with different MIME encodings (Issue that caused white space in front of subject)
```
Comment (Molkobain) [inline test/emailsSample/readme.txt:3]:
```suggestion
email_133_kb4170_multiple_lines_encoded_data.eml: Email with a multiline subject with different MIME encodings (Issue that caused white space in front of subject)
```

Comment (Molkobain) [reply test/emailsSample/readme.txt:3]: Unfortunately as you can see, until now we didn't follow more strict documentation of the email samples test cases. Thanks for adding this up!

**Review by Molkobain — CHANGES_REQUESTED** [review]

**Comment by Molkobain** [comment]
Thanks Jeffrey, I'll try to review it next week so it can go in the bug review, I'll keep you updated.

**Comment by jbostoen** [comment]
@Molkobain is it okay now or are more changes needed?

**Comment by Molkobain** [comment]
Hello Jeffrey,

It should be ok now, I just have to find the time to test/check it locally.

**Comment by Molkobain** [comment]
Sorry for the very long delay Jeffrey, I've just reviewed it and all seem ok to me. I'll bring it up during the next PR review so it can be integrated in the next extension version.

**Review by Molkobain — APPROVED** [review]

**Review by Hipska — APPROVED** [review]

**Comment by Molkobain** [comment]
Accepted in technical review.
@piRGoif volunteered to make the extra step of making it work for other encoding than UTF-8 🙌

**Comment by piRGoif** [comment]
Hello @jbostoen, can you make sure this PR "Allow edits from maintainers" checkbox is enabled ? 
(reference : https://docs.github.com/en/github-ae@latest/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

**Comment by jbostoen** [comment]
@piRGoif  It seems enabled here?

image

**Comment by piRGoif** [comment]
Thanks 👌

**Comment by piRGoif** [comment]
I took quite some time to try to generate an email with another charset in the subject... But didn't achieve to do it :(

Actually there are no charset menu nor preference anymore in Thunderbird (see [this extension](https://addons.thunderbird.net/en-US/thunderbird/addon/charset-menu/) which links to [this bugzilla ticket](https://bugzilla.mozilla.org/show_bug.cgi?id=1704749), and states it was removed in Tb 91 - but it isn't present in [this version release notes](https://www.thunderbird.net/en-US/thunderbird/91.0/releasenotes/) unfortunately :/ )

No more luck with Outlook 2016, Zimbra, Roundcube or GMail.

But... after looking at our existing test base, I found several eml files with multiline subjects having a different charset than utf-8 : 

* test/emailsSample/email_065.eml
* test/emailsSample/email_077.eml
* test/emailsSample/email_107.eml
* test/emailsSample/test gmail.eml

So finally there are some data to play with :)
I'll try to add them to the existing test...

**Comment by piRGoif** [comment]
I've updated the test : the subject value testing is now done in a separate method with its own dataprovider to clear things up.

Among the 6 eml files tested, only the second one (email_133_kb4170_multiple_lines_encoded_data) causes an issue if there are no fix (\RawEmailMessage::DecodeHeaderString).
None of the eml files I've added with non utf-8 encoding are causing a problem...

Considering we don't have any data to reproduce the problem with a non utf-8 encoding, I'm keeping the code as it.

PR is ready for functional review !

**Review by piRGoif — APPROVED** [review]

**Comment by jbostoen** [comment]
@piRGoif Perhaps adjust the regex to match similar patterns. I assume the basic problem will be the same; but for a proper regex I just wanted to make sure there was a proper match.

I notice in your email examples for instance 
`=?ISO-8859-1?`

So I think it would already be safe to adjust the code to not only match UTF-8 but also ISO

**Comment by piRGoif** [comment]
> Perhaps adjust the regex to match similar patterns.

Yes that was my goal in the first place. But as I was unable to produce a similar problem with a non utf-8 charset, I prefer to keep your code. Changing the regexp could lead to drawbacks for no benefit...

**Comment by Molkobain** [comment]
> Yes that was my goal in the first place. But as I was unable to produce a similar problem with a non utf-8 charset, I prefer to keep your code. Changing the regexp could lead to drawbacks for no benefit...

What about adding a comment in the code to explain that we sticked with hardcoded utf-8 on purpose for now because the symptom was not reproduced with the other encoding email sample you tested? That way we won't have those questions when coming back to this portion of code in a few months/years.

**Comment by jbostoen** [comment]
@piRGoif I think you made a mistake in the preg_replace line in the example you posted above. You should not use $1 for the replacement (this would be the first capture group, which is any potential whitespace. It should have been $2.

This lead to incorrect replacements. With $2, your work above seems to work as expected for me.

Only question left is whether the last parameter for iconv_mime_decode should even be $sCharset?
https://www.php.net/manual/en/function.iconv-mime-decode.php 
(  The optional encoding parameter specifies the character set to represent the result by. If omitted or null, iconv.internal_encoding will be used. -> the PHP example at the documentation actually shows a UTF8 string being decoded, and then the resulting string has the ISO-8859-1 character set )

**Comment by piRGoif** [comment]
> I think you made a mistake in the preg_replace line

Yes exatcly, thanks ! I did test with the correct replacement though, but i copy/paste the wrong version of the code, sorry :/

> Only question left is whether the last parameter for iconv_mime_decode should even be $sCharset?

Indeed, the documentation lacks details on this charset parameter... If anyone has inputs on this subject ?

**Comment by Molkobain** [comment]
@jbostoen new Mail To Ticket Automation version should be released by the end of September, don't hesitate to remind me of that PR afterwards 👀

**Comment by jbostoen** [comment]
@Molkobain Reminding ;)

**Comment by Molkobain** [comment]
Hello Jeffrey, I'll confirm with Alexandre that it is ok to merge it in the next version when he'll be back (2nd week of November), reminder set :)

**Comment by Molkobain** [comment]
Hello Jeffrey,

Just discussed this internally, sorry but the next M2T release will include OAuth fixes only to avoid mixing possible regression sources.
This is PR is planned for the release after, or worst case scenario iTop 3.1 release.

**Pull request merged**